### ISSUE TITLE
Reduce the usage of interior mutabilty through out of winit

### ISSUE THREAD

**Issue by kchibisov** [rust-windowing/winit#2697]
This will likely break lots of code downstream in terms of plumbing, and probably will be hell
of annoying for downstream, but I'm getting to the point that it's getting tire some to review
the code in winit due to lots of interior mutability for legit no reason for it to exist in the first place.

For example consider the `Window::set_title(&self, title: String)`. To store the title, I need to add extra `Mutex` just because of the reason to cache it. While in reality it should just be `&mut self`.

Another example is that `Window::new` takes event loop as a `&` instead of `&mut`. And due to that I must add `RefCell` to `Wayland` backend just because of that single call. I had a local patch
to add `&mut` for `run` and `run_return`, but it obviously compiled only on Wayland, because the backend has the most modern architecture among all the backend in winit(because it's being rewritten like each 2 years...).

So what I propose to change:

1. `EventLoop::run` and `EventLoop::run_return` signatures should look like

```rust
 pub fn run<F>(self, event_handler: F) -> !
    where
        F: 'static + FnMut(Event<'_, T>, &mut EventLoopWindowTarget<T>, &mut ControlFlow),
```

See that `EventLoopWindowTarget is now behind `&mut`.

2. Every window call that could change the inner state of the window should take `&mut`. However we might make exceptions in such policy, since some backends will require sync to main thread under the hood, due to winit's nature of window being more like a handle to the original window, and not a real window under the hood.

The 1-st point I'd say mandatory, and 2-nd is nice to have. Because the 1-st is unlikely to create fallout in the ecosystem and will be naturally migrated(because you get object from winit's callback, and pass callback data to the `Window::new`).

The 2-nd point may actually create fallout, because the code relied on the interior mutability, but I think we might force people to write better rust, since interior mutability is a last resort, which is the case in winit actually, because as it was said, the `Window` is a proxy into real window, stored on the main thread.

**Comment by dhardy** [comment]
I tried adapting KAS, and other than needing `DerefMut` for `EventLoop` plus the mentioned change to `run` I ran into no issues compiling. So seems fine to me if `Window` methods take `&mut self` and `Window::new` takes `&mut ELWT`.

**Comment by kchibisov** [comment]
Yeah, the `Window::new` shouldn't be that controversial. It just our backends has issues with it being `&mut`, not the downstream will have issues. For now most of them will require refactoring of the event loop to remove theirs interior infra...

**Comment by notgull** [comment]
Speaking to the two backends I have experience with:

This shouldn't be too difficult in X11, since the only real shared resource is the X11 connection. It'll be a bit harder for Windows, since windows there are actually `!Send` and the current implementation sends callback to the window thread in order to run things, not to mention there's a bit more synchronization in the first place.

**Comment by kchibisov** [comment]
I'm not saying to completely remove. I'm saying that public API winit has shouldn't enforce interior mutability on us. Like for example `set_title` on `Window` where you simple need to cache it on the `Window` to do a getter later on.

**Comment by maroider** [comment]
1. `&mut EventLoopWindowTarget<T>` seems like it wouldn't be all that problematic. Technically this would allow someone to use `mem::swap` on it, but that's a very silly thing to do, and winit would likely just panic.
2. I'm not sold on this idea. In practical terms, this is something downstream can likely adapt to, but I don't feel it's a necessary change. In my opinion, interior mutability is a perfectly acceptable thing to use for a cache, and I don't think code which implements a cache without interior mutability is necessarily better code.

**Comment by kchibisov** [comment]
>&mut EventLoopWindowTarget<T> seems like it wouldn't be all that problematic. Technically this would allow someone to use mem::swap on it, but that's a very silly thing to do, and winit would likely just panic.

You can't create multiple event loops in winit, so you can't get multiple targets.

**Comment by maroider** [comment]
Ah, right, I forgot that we added a check for that.

**Comment by madsmtm** [comment]
I understand the frustration, but I think we should consider the opposite: That `winit` owns the window data, and exposes a handle to it instead. So e.g. change `Window` to a clone-able `WindowHandle` (that is maybe `Send + Sync`, maybe locked to the UI thread, not sure). I know, windows are currently closed on `Drop`, but that is IMO the wrong API since windows can be closed by the system at any time anyhow.

This would also open the door for something like `EventLoopWindowTarget::windows(&self) -> Vec<WindowHandle>`, and using `WindowHandle` instead of `WindowId` in `WindowEvent`, both of which I think could be quite nice.

---

I know that this is basically a tug-of-war between what is most convenient in each particular backend, though I also suspect that the current design (only one handle pr. window allowed, the window is closed on `Drop`) is a bit restrictive.

**Comment by kchibisov** [comment]
>I know, windows are currently closed on Drop, but that is IMO the wrong API since windows can be closed by the system at any time anyhow.

That's not true, you're asked to close, if your system closes at any time(besides pkill), the system is wrong. Every protocol requests close, but you might ignore.

>This would also open the door for something like EventLoopWindowTarget::windows(&self) -> Vec<WindowHandle>, and using WindowHandle instead of WindowId in WindowEvent, both of which I think could be quite nice.

Nah, the current API is fine, the `Window` can't be `Clone`, the `Handle` can't be clone as well from what I know. Some platforms allow access to window from multiple threads (Wayland).

My main issue is window creation, since due to `&` I have to use `RefCell` in Wayland backend, which I hate, because there's no need for it in the first place.

**Comment by madsmtm** [comment]
> My main issue is window creation

I'd be fine with `&mut EventLoopWindowTarget<T>`.

> That's not true, you're asked to close, if your system closes at any time(besides pkill), the system is wrong. Every protocol requests close, but you might ignore.

Hmm, yeah, I can see now that that's true. And I was going to argue about windows being created by the system, but that requires special support anyhow (e.g. `WindowEvent::NewRequested` or similar), and could also be ignored.

Still, I think it's a bit of a hassle that you have to keep the list of windows alive yourself.

**Comment by kchibisov** [comment]
>Still, I think it's a bit of a hassle that you have to keep the list of windows alive yourself.

It's not though, because you can move `Window` to a different thread, so you won't have a list.

**Comment by dhardy** [comment]
> Still, I think it's a bit of a hassle that you have to keep the list of windows alive yourself.

I don't. Any non-empty window has other data associated with it; it's no problem storing `winit::window::Window` in the same struct.

**Comment by rib** [comment]
It seems worth noting that when discussing this topic a bit further on IRC it did also start to look like the implementation of `Window::new()` on Wayland has an awkward design currently that requires a server round trip which also dispatches Wayland callbacks (and would lead to a callback re-entrancy issue if it wasn't also buffering events internally).

This is notable because one of the noted reasons for wanting the `&mut ELWT` was for the sake of being able to pass that through to `calloop` in the Wayland backend when dealing with this synchronous round trip.

I don't have much of an opinion on whether it would be better to pass around an `&mut ELWT` (it generally seems like that wouldn't be a huge change) but I do think there's probably a bigger design issue to resolve in the Wayland backend currently, and this issue might partly be a symptom of that.

It looks like the Wayland backend needs one of these:
1. A way to complete the creation of a Window asynchronously - such as a `WindowReady` event. Trying to synchronously run the event loop inside `Window::new` looks like a big design problem imho. If creation _really_ requires an async response from the server (though I currently also find this a bit suspicious) then that should almost certainly be driven via the actual event loop.
3. A new, completely async API to create a window in the first place (so instead of getting a `Window` that's not initially usable, you get an event that hands you a `Window` when it's ready.
4. To re-evaluate what's wrong with handing out a `Window` before getting back a `wl_configure` event. I really tend to think that shouldn't be strictly necessary. At least checking old code I have for creating wayland windows, surfaces and EGLSurfaces I didn't need a sync roundtrip and conceptually I just don't really see why that would be necessary (the fact the compositor might pick a different size shouldn't be a big problem as far as I know)

It was noted on IRC that:
```
On Wayland, the window is ready once you've got a `configure` for your role surface.
If you submit a buffer before that, you'll get a protocol error.
```
but I'm not sure that's a full story for why a sync roundtrip is needed.

**Comment by kchibisov** [comment]
>but I'm not sure that's a full story for why a sync roundtrip is needed.

https://wayland.app/protocols/xdg-shell#xdg_surface 

For fwiw the current sync callback dispatch doesn't cause any issues, given that every callback event is batched, so you don't have callback stacking issues.

But in general, Wayland is async and any `sync` modeled request in winit could result in callback stacking.

**Comment by rib** [comment]
The backend should always be dispatching from a controlled point in the event loop so there shouldn't generally be a concern with reentrancy / callback stacking with wayland. There are some special cases where e.g. egl sync with the compositor for swap buffers but I'm pretty sure that wont dispatch other callbacks.

> For fwiw the current sync callback dispatch doesn't cause any issues, given that every callback event is batched, so you don't have callback stacking issues.

yeah, but the fact that the backend is even going this buffering at all just seems like a sign that there's something amiss here, that shouldn't be needed.
